### DISCUSSION TITLE
How to implement data fetching logic in suspended function with Spring Webflux?

### DISCUSSION THREAD

**Discussion by Infergo** [ExpediaGroup/graphql-kotlin#971] [General] [answered]
I am trying to implement a relation between two entities.
I understand from the [documentation](https://expediagroup.github.io/graphql-kotlin/docs/schema-generator/execution/fetching-data) that the best way to do this is via a `suspend fun`.
However, the documentation then shows a comment with "logic to fetch reviews here".
Exactly that part is where I'm stuck.
I'd like to use a repository but Spring can not inject this, and I have no way of instantiating a repository.

I have a Query controller with a specific schema class, and a simple repository.
Is there a way to use the repository and grab the companies from the database?
Please let me know if I need to provide more info.

```
@Component
class UserQueryController : Query {
    fun user(keycloakId: String): CompletableFuture<UserClassSchema> {
        return userRepository.findByKeycloakId(keycloakId)
            .map { u ->
                UserClassSchema(
                    id = u.id!!,
                    keycloakId = u.keycloakId,
                    email = u.email,
                    firstName = u.firstName,
                    lastName = u.lastName
                )
            }
            .toFuture()
    }
}
```

```
@GraphQLName("User")
data class UserClassSchema(
    val id: String,
    val keycloakId: String,
    val email: String,
    val firstName: String,
    val lastName: String
) {
    suspend fun companies(): List<Company> {
        // Repository usage here?
    }
}
```

```
interface CompanyRepository : ReactiveCrudRepository<Company, String>
```

**Comment by dariuszkuc** [comment, marked as answer]
Hello 👋 
There are number of ways this can implemented. If you are using latest `graphql-kotlin-spring-server` i.e. 4.x (currently alpha) release we default to use [SpringDataFetcher](https://expediagroup.github.io/graphql-kotlin/docs/spring-server/spring-schema#spring-data-fetcher) that allows you to autowire Spring beans to method arguments.

```kotlin
@Component
class UserQueryController(private val userRepository: UserRepository) : Query {
    suspend fun user(keycloakId: String): UserClassSchema {
        return userRepository.findByKeycloakId(keycloakId)
            .map { u ->
                UserClassSchema(
                    id = u.id!!,
                    keycloakId = u.keycloakId,
                    email = u.email,
                    firstName = u.firstName,
                    lastName = u.lastName
                )
            }
            .await()
    }
}

@GraphQLName("User")
class UserClassSchema(
    val id: String,
    val keycloakId: String,
    val email: String,
    val firstName: String,
    val lastName: String
) {
    suspend fun companies(@GraphQLIgnore @Autowired repository: CompanyRepository): List<Company> {
        return repository.retrieveCompany()
    }
}

```

**Reply by smyrick** [reply]
I will add to this that you can also pass the repository class to your schema class and have it not added through DI but just as a private constructor arg. Then you can use it in functions not expose anything in the schema

```kotlin
class UserClassSchema(
    val id: String,
    val keycloakId: String,
    val email: String,
    val firstName: String,
    val lastName: String,
    private val repository: CompanyRepository
) {
    suspend fun companies(): List<Company> {
        return repository.retrieveCompany(id)
    }
}
```

**Comment by Infergo** [comment]
Thanks guys, I got it working with your suggestions!
